File: frontend/neuraapp/NeuraApp/Services/ApiService.swift
import Foundation

// Custom Errors remain the same.
enum APIError: Error, LocalizedError {
    case invalidURL
    case invalidResponse
    case requestFailed(description: String)
    case decodingError

    var errorDescription: String? {
        switch self {
        case .invalidURL: return "The server URL is invalid."
        case .invalidResponse: return "The server returned an invalid response."
        case .requestFailed(let description): return description
        case .decodingError: return "There was an error decoding the server's response."
        }
    }
}

class ApiService {
    static let shared = ApiService()
    private let keychainService = KeychainService.shared

    private let jsonDecoder: JSONDecoder = {
        let decoder = JSONDecoder()
        
        // 1. Create a Date Formatter that understands fractional seconds.
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSSSS" // This MUST match your backend's format
        formatter.calendar = Calendar(identifier: .iso8601)
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        formatter.locale = Locale(identifier: "en_US_POSIX")

        // 2. Tell the decoder to use our custom formatter.
        decoder.dateDecodingStrategy = .formatted(formatter)
        
        return decoder
    }()
    
    // MARK: - Module 1: UserHub

    func login(email: String, password: String) async throws -> String {
        guard let url = URL(string: "\(Constants.userHubBaseUrl)/auth/token") else {
            throw APIError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        
        var components = URLComponents()
        components.queryItems = [
            URLQueryItem(name: "username", value: email),
            URLQueryItem(name: "password", value: password)
        ]
        request.httpBody = components.query?.data(using: .utf8)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw APIError.requestFailed(description: "Invalid email or password.")
        }
        
        do {
            let loginResponse = try jsonDecoder.decode(LoginResponse.self, from: data)
            return loginResponse.accessToken
        } catch {
            throw APIError.decodingError
        }
    }

    func getMyProfile() async throws -> User {
        guard let url = URL(string: "\(Constants.userHubBaseUrl)/users/me") else {
            throw APIError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        guard let token = keychainService.getToken() else {
            throw APIError.requestFailed(description: "User is not authenticated.")
        }
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw APIError.invalidResponse
        }
        
        do {
            let user = try jsonDecoder.decode(User.self, from: data)
            return user
        } catch {
            throw APIError.decodingError
        }
    }

    func performCheckIn(forUser user: User) async throws {
        guard let url = URL(string: "\(Constants.userHubBaseUrl)/attendance/check-in") else {
            throw APIError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        guard let token = keychainService.getToken() else {
            throw APIError.requestFailed(description: "User is not authenticated.")
        }
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body = ["user_id": user.id, "location": "NeuraApp Mobile Login"] as [String: Any]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        let (_, response) = try await URLSession.shared.data(for: request)
        if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode != 200 {
            print("Check-in failed with status code: \(httpResponse.statusCode)")
        }
    }

    // MARK: - Module 2: NeuraNLP Agent
    
    func submitQuery(_ query: String) async throws -> String {
        guard let url = URL(string: "\(Constants.nlpAgentBaseUrl)/query") else {
            throw APIError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        guard let token = keychainService.getToken() else {
            throw APIError.requestFailed(description: "User is not authenticated.")
        }
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let boundary = "Boundary-\(UUID().uuidString)"
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"query\"\r\n\r\n".data(using: .utf8)!)
        body.append("\(query)\r\n".data(using: .utf8)!)
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"mode\"\r\n\r\n".data(using: .utf8)!)
        body.append("text\r\n".data(using: .utf8)!)
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)
        request.httpBody = body
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw APIError.invalidResponse
        }
        
        struct NLPResponse: Codable { let response: String }

        do {
            let nlpResponse = try jsonDecoder.decode(NLPResponse.self, from: data)
            return nlpResponse.response
        } catch {
            throw APIError.decodingError
        }
    }
}

File: frontend/neuraapp/lib/services/api_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:neuraapp/models/user_model.dart';
import '../utils/constants.dart';
import 'secure_storage_service.dart';

class ApiService {
  final SecureStorageService _storageService;

  ApiService(this._storageService);

  // Helper method for headers requiring authorization
  Future<Map<String, String>> _getAuthHeaders() async {
    final token = await _storageService.getToken();
    if (token == null) throw Exception('User not authenticated.');
    return {
      'Authorization': 'Bearer $token',
    };
  }

  // Endpoint 1.1: User Login
  Future<String> login(String email, String password) async {
    final response = await http.post(
      Uri.parse('${AppConstants.userHubBaseUrl}/auth/token'),
      headers: {'Content-Type': 'application/x-www-form-urlencoded'},
      body: {'username': email, 'password': password},
    );

    if (response.statusCode == 200) {
      return jsonDecode(response.body)['access_token'];
    } else {
      throw Exception('Invalid email or password.');
    }
  }

  // Endpoint 1.2: Get My Profile
  Future<User> getMyProfile() async {
    final headers = await _getAuthHeaders();
    final response = await http.get(
      Uri.parse('${AppConstants.userHubBaseUrl}/users/me'),
      headers: headers,
    );

    if (response.statusCode == 200) {
      return User.fromJson(jsonDecode(response.body));
    } else {
      throw Exception('Failed to load user profile. Your session may have expired.');
    }
  }

  // Endpoint 2.1: Submit a Query
  Future<String> submitQuery(String query) async {
    final headers = await _getAuthHeaders();
    var request = http.MultipartRequest(
      'POST',
      Uri.parse('${AppConstants.nlpAgentBaseUrl}/query'),
    );
    request.headers.addAll(headers);
    request.fields['query'] = query;
    request.fields['mode'] = 'text';

    var streamedResponse = await request.send();
    var response = await http.Response.fromStream(streamedResponse);

    if (response.statusCode == 200) {
      return jsonDecode(response.body)['response'];
    } else {
      throw Exception('Failed to get response from AI agent.');
    }
  }

  // Endpoint 1.3: Check-In/Out
  Future<void> performCheckIn(int userId) async {
    final headers = await _getAuthHeaders();
    headers['Content-Type'] = 'application/json; charset=UTF-8';

    final body = jsonEncode({
      'user_id': userId,
      'location': 'NeuraApp Mobile Login',
    });

    try {
      await http.post(
        Uri.parse('${AppConstants.userHubBaseUrl}/attendance/check-in'),
        headers: headers,
        body: body,
      );
    } catch (e) {
      print("Check-in failed but not blocking user: $e");
    }
  }

  // THIS IS THE METHOD THAT WAS MISSING.
  // Purpose: Send the Firebase Cloud Messaging (FCM) token to the backend.
  Future<void> registerDevice(String fcmToken) async {
    try {
      final headers = await _getAuthHeaders();
      headers['Content-Type'] = 'application/json; charset=UTF-8';
      
      await http.post(
        // NOTE: This uses the future endpoint from the brief.
        Uri.parse('${AppConstants.userHubBaseUrl}/users/me/register-device'),
        headers: headers,
        body: jsonEncode(<String, String>{'fcm_token': fcmToken}),
      );
      print("FCM Token successfully registered with backend.");
    } catch (e) {
      print("Failed to register FCM token with backend: $e");
      // Not throwing an error as this is not a critical user-facing failure.
    }
  }
}